### ISSUE TITLE
5.6 regression: Incorrect param type inference for type with all optional props

### ISSUE THREAD

**Issue by trevorade** [microsoft/TypeScript#59656]
### 🔎 Search Terms

inference, param

### 🕗 Version & Regression Information

- This changed between versions 5.5.4 and 5.6.0

### ⏯ Playground Link

https://www.typescriptlang.org/play/?ts=5.6.0-dev.20240816#code/JYOwLgpgTgZghgYwgAgMoHsC2EAqBPABxQG8AoZC5AeiuQHkRkBWAOgDYWAGAGmQgA8iUYNnDIARnmSY4Aa1ABzZGAAWwAM7ICUdAWRQIARwCuwAwBMW5SsZDATEAAo6CAfgBcydWGEgFAblIAX1JScwgEABs4A2QEdBBvZHRxdU86VOgANzhxSIgAHgxsfCIAPkDSGFsEMGAE5BiFAHEIMHUAdR0-UogACgBKdMyoHLzC4iCy5DJKfTbjKEYU9RYCYCI+sDgCPqbkAF5p2bnKJsDTyhoAPVdrS5pkAFFBCMhzT2LcQhRQZhZWAAWe6nR4AQVqxjgkU8FEmyD+WVYHE4AFpxG04Mp0P8UajwlkWAAmThEwGcAAcAEY2Mg9mBkPk4N4BiCggMBoEQlUanUGk1Wu0AMLoKAGWq9QYzEEGMCLZapNYbfrbXb7I7Sy4Uc4guY3O5a6i0MGRADucDwmniYrebI5XNCoEgsEQKAy6myuXyBRwxxB6yIBTBZT6unSQjgYFFADFefUQD7eMGhvQRmNvcGHU7oPAkMgAKogGJ4WMgWrxxPIABKxzp6nQiyQnhwKar-mQIWzLrzdAjUagpfLCUrNb4-EgIHMmkLxcHfIT7s94x9ZV4i9GXsKNdrnfAOddyAAsgl0L1e9BIzG48PfWOJ1P6H2r2X55Xb8QO2EItFYtUX-HlB2Fc+hAAQwE8PockiYwIGbAZDmmLJ0GAcwU2PEBTx+c8oEvAdrwTX1AiAA

### 💻 Code

```ts
interface SomeType {
    // On 5.6.0, experiment by making this prop required.
    uniqueProp?: string;
}

declare const obs: Observable<SomeType>;

function argGetsWrongType(): Observable<{}> {
    return obs.pipe(tap(arg => {
        arg;
        //^?
        // Expected: SomeType in 5.5.4
        // Actual:   {} in v5.6.0-beta to 5.6.0-dev.20240816 (at least)
    }));
}

function argGetsCorrectType() {
    return obs.pipe(tap(arg => {
        arg;
        //^?
        // Always correct
    }));
}

interface Observable<T> {
    pipe<A>(op: OperatorFunction<T, A>): Observable<A>;
}
interface UnaryFunction<T, R> { (source: T): R; }
interface OperatorFunction<T, R> extends UnaryFunction<Observable<T>, Observable<R>> { }
interface MonoTypeOperatorFunction<T> extends OperatorFunction<T, T> { }
declare function tap<T>(next: (value: T) => void): MonoTypeOperatorFunction<T>;
```

### 🙁 Actual behavior

`arg` is inferred as `{}`

### 🙂 Expected behavior

`arg` is inferred as `SomeType` as was consistent in v5.5.4

### Additional information about the issue

(Google note: See http://cl/664889390 for local workaround)

**Comment by Andarist** [comment]
Bisects to [this diff](https://github.com/microsoft/TypeScript/compare/b1c52c53cc6c8cf35b19accfc3e29916489c821e...fb88f027b015d95816e91f6b4f0e67dc57fa7a9f), so the change likely was caused by https://github.com/microsoft/TypeScript/pull/57909

**Comment by trevorade** [comment]
FWIW, in Google's codebase, I'm seeing other new inference related bugs popping up. For the life of me, I can't make a simplified repo for this particular case...

I'll try to add other repros when I can...

**Comment by Andarist** [comment]
I’d appreciate any kind of repros - they could even be not-so-minimal 😉

**Comment by RyanCavanaugh** [comment]
A useful observation is that this can be fixed by a single variance annotation
```ts
interface Observable<in T> {
```
which implies we're likely measuring the variance wrong.

**Comment by trevorade** [comment]
> I’d appreciate any kind of repros - they could even be not-so-minimal 😉

Yeah. This one is just some internal team who has produced a pipeline with some pretty complicated TS types and inference is breaking in a new way. I'll be looking more at overall compilation issues in the codebase and will try to highlight other similar issues when they seem relevant.

**Comment by trevorade** [comment]
By the way, I've identified a number of other new compilation issues a little different from this that I believe is related to this issue.

Once there is a new 5.6.0 build with this PR, I can see if it resolves the other new issues I've been seeing.

In this example, basically, inference appears to be dropping optional parameters from a type or making them required. I haven't dug into it super deep. If this PR doesn't fix the issue, I'll try to boil it down to a repro.

**Comment by jakebailey** [comment]
Can you try the build on https://github.com/microsoft/TypeScript/pull/59709#issuecomment-2305582124 to see if that fixes your issues?

**Comment by trevorade** [comment]
Hey there. Confirmed that this fix does fix the specific issue noted here.

That said, I am still seeing other new inference-related bugs that I was hoping this fix would resolve. I'll try to construct repros in TS Playground for those ones and file new issues.
